sdk/identity/Azure.Identity/src/AzureCliCredential.cs

```diff
@@ -0,0 +1,142 @@
+        /// <param name="requestContext"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken = default)
+        {
+            return GetTokenImplAsync(requestContext, cancellationToken).GetAwaiter().GetResult().GetTokenOrThrow();
+        }
+
+        /// <summary>
+        /// Obtains a access token from Azure CLI service, using the access token to authenticate. This method id called by Azure SDK clients.
+        /// </summary>
+        /// <param name="requestContext"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken = default)
+        {
+            return (await GetTokenImplAsync(requestContext, cancellationToken).ConfigureAwait(false)).GetTokenOrThrow();
+        }
+
+        async ValueTask<ExtendedAccessToken> IExtendedTokenCredential.GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            return await GetTokenImplAsync(requestContext, cancellationToken).ConfigureAwait(false);
+        }
+
+        ExtendedAccessToken IExtendedTokenCredential.GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            return GetTokenImplAsync(requestContext, cancellationToken).GetAwaiter().GetResult();
+        }
+
+        private async ValueTask<ExtendedAccessToken> GetTokenImplAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            using CredentialDiagnosticScope scope = _pipeline.StartGetTokenScope("Azure.Identity.AzureCliCredential.GetToken", requestContext);
+
+            try
+            {
+                string resource = ScopeUtilities.ScopesToResource(requestContext.Scopes);
+                string resourcePatter = "^[0-9a-zA-Z-.:/]+$";
+                bool isResourceMatch = Regex.IsMatch(resource, resourcePatter);
+
+                if (!isResourceMatch)
+                {
+                    throw new Exception(InvalidResourceMessage);
+                }
+
+                (string output, int exitCode) = _client.GetAzureCliAccesToken(resource);
+
+                if (exitCode != 0)
+                {
+                    bool isLoginError = output.StartsWith("Please run 'az login'", StringComparison.CurrentCultureIgnoreCase);
+                    bool isWinError = output.StartsWith(WinAzureCLIError, StringComparison.CurrentCultureIgnoreCase);
+                    string pattter = "az:(.*)not found";
+                    bool isOtherOsError = Regex.IsMatch(output, pattter);
+
+                    if (isWinError || isOtherOsError)
+                    {
+                        throw new AuthenticationFailedException(AzureCLINotInstalled);
```

Comment:
Sorry I missed the question by @ellismg earlier. I agree this should be a `CredentialUnavailableException`,

Reply:
Also the `IExtendedTokenCredential` interface is meant to provide an abstraction where exceptions are returned rather than thrown. This allows the caller to determin if the error should be raised.

Reply:
Detail in this [PR](https://github.com/wantedfast/azure-sdk-for-net/pull/1)